### PR TITLE
Feature/improved joystick

### PR DESCRIPTION
**Pull request by drowhunter**
1. Fixed Issue where Globals with [NeedIndexer] attribute would fail to parse, if script contained code with similar function name

example 
```
diagnostics.watch(a)
diagnostics.watchObject(a)  # (bug) not allowed because the pre-parser matched only on the word 
```
 

2. Improved Joystick Globals
Joystick axes are now producing values between -1 and 1, and can be configured as a half axis, inverted or any combination of both

Example 
```
if starting:
    joystick[0].config.rotationX = axisConfig.HalfAxisInverted # value will be between 1 and 0
    joystick[0].config.rotationX = axisConfig.HalfAxis # value will be between 0 and 1
    joystick[0].config.rotationX = axisConfig.FullAxisInverted # value will be between 1 and -1
    joystick[0].config.rotationX = axisConfig.FullAxis # value will be between -1 and 1 (Default)
    joystick[0].config.rotationX = axisConfig.Raw # value will be between 0 - 65535

diagnostics.watch(joystick[0].rotationX) 
```

3. Fixed the `ensureMaprange` function.  
`ensureMapRange` function did not work if you flipped the output to invert it's output

### CODE REVIEW

File: FreePIE.Core.Plugins/joystick/JoystickGlobal.cs [hunk 1]
```diff
@@ -13,152 +13,193 @@ public class JoyProp
...
+    {
+        public bool up { get; set; }
+        public bool down { get; set; }
+        public bool left { get; set; }
+        public bool right { get; set; }
 
-        internal int numAxes { get { return _joystick.Capabilities.AxeCount; } }
 
-        internal int numPovs { get { return _joystick.Capabilities.PovCount; } }
+        //public DpadGlobal() {  }
```
Comment (Ofisare) [inline FreePIE.Core.Plugins/joystick/JoystickGlobal.cs]: if not required remove

File: FreePIE.Core.Plugins/joystick/JoystickGlobal.cs [hunk 2]
```diff
@@ -13,152 +13,193 @@ public class JoyProp
...
-        internal Dictionary<JoystickOffset, JoyProp> AxisInfo { get; }
+        //internal void update(int pov)
+        public DpadGlobal(int pov = -1)
+        {
+            if (pov >= 0)
+            {
+                if(pov >= 360)
+                    pov = pov % 360;
 
+                up = pov == 0 || pov == 45;
```
Comment (Ofisare) [inline FreePIE.Core.Plugins/joystick/JoystickGlobal.cs]: that should be up = pov >= 315 || pov <= 45 etc for the other cases, if you allow int, you should consider all values, and why int, not double?

File: FreePIE.Core.Plugins/joystick/JoystickGlobal.cs [hunk 3]
```diff
@@ -13,152 +13,193 @@ public class JoyProp
...
+                down = pov == 135 || pov == 180;
```
Comment (Ofisare) [inline FreePIE.Core.Plugins/joystick/JoystickGlobal.cs]: btw. there would be 3 cases, down, down left and down right ;-)

File: FreePIE.Core/ScriptEngine/Globals/ScriptHelpers/DiagnosticHelper.cs [hunk 4]
```diff
@@ -31,39 +33,66 @@ public void debug(object arg)
         [NeedIndexer]
         public void watch(object value, string indexer)
         {
+            
+            if(value != null && !IsSimple(value.GetType()))
```
Comment (Ofisare) [inline FreePIE.Core/ScriptEngine/Globals/ScriptHelpers/DiagnosticHelper.cs]: empty space after if :-)

File: FreePIE.Core.Plugins/joystick/JoystickPlugin.cs [hunk 5]
```diff
@@ -99,7 +99,7 @@ public override void Stop()
                 }
                 finally
                 {
-                    j.Dispose();
+                    ((IDisposable)j).Dispose();
```
Comment (Ofisare) [inline FreePIE.Core.Plugins/joystick/JoystickPlugin.cs]: unrequired

File: FreePIE.Core.Plugins/joystick/JoystickGlobal.cs [hunk 6]
```diff
@@ -13,152 +13,193 @@ public class JoyProp
...
-            //    return ensureMapRange(value, p.props.Range.Minimum, p.props.Range.Maximum, -1, 1);
-            //}
-            return value;
+            return Math.Max(Math.Min(mapRange(x, xMin, xMax, yMin, yMax), Math.Max(yMin, yMax)), Math.Min(yMin, yMax));
+        }
+
+        
+        private double normalize(int value, axisConfig cfg )
+        {
+            if(cfg == axisConfig.Raw)
```
Comment (Ofisare) [inline FreePIE.Core.Plugins/joystick/JoystickGlobal.cs]: empty space after if :-)

File: FreePIE.Core.Plugins/joystick/JoystickGlobal.cs [hunk 7]
```diff
@@ -13,152 +13,193 @@ public class JoyProp
...
+
+        public DpadGlobal[] dpad => pov.Select(p => new DpadGlobal(p)).ToArray();
+
+
+
+
+        #endregion
+
+        #region Private Methods
+        private double mapRange(double x, double xMin, double xMax, double yMin, double yMax)
```
Comment (Ofisare) [inline FreePIE.Core.Plugins/joystick/JoystickGlobal.cs]: make the FilterHelper functions static so you can reuse them

File: FreePIE.Core.Plugins/joystick/JoystickGlobal.cs [hunk 8]
```diff
@@ -13,152 +13,193 @@ public class JoyProp
...
 
+    [Global(Name = "joystick")]
+    public class JoystickGlobal : IDisposable
+    {
 
+        private readonly Joystick _joystick;
+
+        private JoystickState _state;
+        
+        //public JoystickState state => _state;
```
Comment (Ofisare) [inline FreePIE.Core.Plugins/joystick/JoystickGlobal.cs]: better to clean up if not needed ;-)

File: FreePIE.Core.Plugins/joystick/JoystickGlobal.cs [hunk 9]
```diff
@@ -13,152 +13,193 @@ public class JoyProp
...
+        
+        private Dictionary<axisConfig, (int min, int max)> _axisConfigValue;
```
Comment (Ofisare) [inline FreePIE.Core.Plugins/joystick/JoystickGlobal.cs]: you are not modifying this, might be better to make this a static property with a static initializer

File: FreePIE.Core.Plugins/joystick/JoystickGlobal.cs [hunk 10]
```diff
@@ -13,152 +13,193 @@ public class JoyProp
...
+        
 
+        #region Visible Global Properties
 
+        public string name { get { return _joystick?.Information.ProductName ?? ""; } }
 
-        public float x
-        {
-            get 
+        public JoyCount count => new JoyCount
```
Comment (Ofisare) [inline FreePIE.Core.Plugins/joystick/JoystickGlobal.cs]: this should be done in the constructor to reduce uneccessary (de)allocations

File: FreePIE.Core.Plugins/joystick/JoystickGlobal.cs [hunk 11]
```diff
@@ -13,152 +13,193 @@ public class JoyProp
...
+        public double rotationX => normalize(_state?.RotationX ?? 0, config.rotationX);
+        public double rotationY => normalize(_state?.RotationY ?? 0, config.rotationY);
+        public double rotationZ => normalize(_state?.RotationZ ?? 0, config.rotationZ); 
+            
         
 
         public bool[] buttons { get; private set; }
 
-        public float[] sliders
+        public double[] sliders
```
Comment (Ofisare) [inline FreePIE.Core.Plugins/joystick/JoystickGlobal.cs]: these two are also expensive, you don't want to trigger the garbage collection ingame :D
maybe we should use slider(int index) and pov(index) instead of creating arrays all the time

File: FreePIE.Core.Plugins/joystick/JoystickGlobal.cs [hunk 12]
```diff
@@ -13,152 +13,193 @@ public class JoyProp
...
         }
 
         public int[] pov
         {
             get
             {
-                return numPovs > 0 ? State.PointOfViewControllers.Select(p =>
+                return _joystick.Capabilities.PovCount > 0 ? _state.PointOfViewControllers.Take(count.povs).Select(p =>
                 {
                     var r = p > 0 ? p / 100 : p;
```
Comment (Ofisare) [inline FreePIE.Core.Plugins/joystick/JoystickGlobal.cs]: what is coming out here? a value between 0 and 1 or 0 and 360? we should write a comment here for interested readers to not have to guess

**Review by Ofisare — CHANGES_REQUESTED** [review]

File: FreePIE.Core.Plugins/joystick/JoystickGlobal.cs (hunk 2 above)
Comment (drowhunter) [reply FreePIE.Core.Plugins/joystick/JoystickGlobal.cs:66]: From Copilot.
```
in DirectInput, the values for a Point of View (POV) hat switch are typically represented by 8 directions. These directions correspond to the cardinal and intercardinal points: North, North-East, East, South-East, South, South-West, West, and North-West.

the values returned for the POV (Point of View) hat switch in DirectInput are typically integers. These values represent the angles of the hat switch position in degrees, where:

0 represents North
4500 represents North-East
9000 represents East
13500 represents South-East
18000 represents South
22500 represents South-West
27000 represents West
31500 represents North-West

When the hat switch is in the center (neutral position), the value is usually -1.
```
This also lines up with what i have noticed. you will never receive a value for a pov hat other than what is listed above.

File: FreePIE.Core.Plugins/joystick/JoystickGlobal.cs (hunk 3 above)
Comment (drowhunter) [reply FreePIE.Core.Plugins/joystick/JoystickGlobal.cs:67]: see above

File: FreePIE.Core.Plugins/joystick/JoystickGlobal.cs (hunk 2 above)
Comment (drowhunter) [reply FreePIE.Core.Plugins/joystick/JoystickGlobal.cs]: yeah the values are incorrect,  ill fix that.  as for why int.  

From Microsoft.  There are only 9 values you will ever get back.


the values returned for the POV (Point of View) hat switch in DirectInput are typically integers. These values represent the angles of the hat switch position in degrees, where:

0 represents North

4500 represents North-East

9000 represents East

13500 represents South-East

18000 represents South

22500 represents South-West

27000 represents West

31500 represents North-West

When the hat switch is in the center (neutral position), the value is usually -1.

File: FreePIE.Core.Plugins/joystick/JoystickGlobal.cs (hunk 3 above)
Comment (drowhunter) [reply FreePIE.Core.Plugins/joystick/JoystickGlobal.cs]: yes, i will correct the values,   the idea is that if you are pushing the diagonals then two of the booleans will be true.  e.g northeast would be up = true and right = true

File: FreePIE.Core.Plugins/joystick/JoystickPlugin.cs (hunk 5 above)
Comment (drowhunter) [reply FreePIE.Core.Plugins/joystick/JoystickPlugin.cs:102]: not required to dispose the joysticks?

File: FreePIE.Core.Plugins/joystick/JoystickGlobal.cs (hunk 7 above)
Comment (drowhunter) [reply FreePIE.Core.Plugins/joystick/JoystickGlobal.cs]: actually this is a weird one.  I wanted to just use the function from Freepie.Core FilterHelper 
but plugins do not reference Freepie.Core.  So i had to copy it over here for private use,  not meant to be reused.

I could add a reference to Freepie.Core from plugins,  but that would go against what they were going for with the plugin system. (It only references Contracts, so that plugin authors can make their own plugins and just drop them in the freepie folder)

So Long story short these methods are private and will not be reused.

Comment (drowhunter) [reply FreePIE.Core.Plugins/joystick/JoystickGlobal.cs]: Unless you mean to allow this one function to be reused somewhere else within Freepie.Plugin ,  in which case we would be better off creating an interface for FilterHelper and adding it into FreePie contracts. 

That might not work either since Filterhelper isnt actually registered with the Container directly.  All of the Helpers are simply registered as IScriptHelper and resolve as an IEnumerable.  

So we would need to individually register IFilterHelper with the container.   

It gets really complicated.

**Comment by drowhunter** [comment]
Resolved all but one
